### PR TITLE
JIT: Convert some of the old-style intrinsics to NamedIntrinsics

### PR DESCRIPTION
**Pull request by EgorBo**
It's needed for inliner that handles only NamedIntrinsics. Eventually, I guess, we'll get rid of all of the old-style ones.

### CODE REVIEW

**Comment by EgorBo** [comment]
@dotnet/jit-contrib PTAL

**Comment by EgorBo** [comment]
@dotnet/jit-contrib @echesakov PING 🙂 I'd love to merge it and unblock my next inliner-related PR on top of it

File: src/coreclr/jit/gentree.cpp [hunk 1]
```diff
@@ -11836,15 +11834,7 @@ void Compiler::gtDispTree(GenTree*     tree,
...
-                switch (intrinsic->gtIntrinsicId)
-                {
-                    case CORINFO_INTRINSIC_Object_GetType:
-                        printf(" objGetType");
-                        break;
-
-                    default:
-                        unreached();
-                }
+                unreached();
```
Comment (echesakovMSFT) [inline src/coreclr/jit/gentree.cpp:11837]: This should either  `unreached()` or `assert(intrinsic->gtIntrinsicName == NI_Illegal)` ?

File: src/coreclr/jit/valuenum.cpp [hunk 2]
```diff
@@ -8753,9 +8753,9 @@ void Compiler::fgValueNumberIntrinsic(GenTree* tree)
     }
     else
     {
-        switch (intrinsic->gtIntrinsicId)
+        switch (intrinsic->gtIntrinsicName)
```
Comment (echesakovMSFT) [inline src/coreclr/jit/valuenum.cpp:8756]: Seems weird to have `switch` with just one `case` and default asserting with `unreached()`. 
Should this be replaced with `assert(intrinsic->gtIntrinsicName == NI_System_Object_GetType)` ?

**Review by echesakov — APPROVED** [review]
Sorry for keep you waiting..
Overall, looks good - left some comments.

BTW, have you encountered any diffs with this change?

**Comment by AndyAyersMS** [comment]
Let me echo what @echesakov wrote above: you really need to show that this doesn't lead to diffs. Given the the changes in the jit interface you can't use SPMI -- you should use  jit diffs in both PMI and Crossgen2 modes for this, running base and diff separately (since the jits can't be swapped), over all the frameworks assemblies.

**Comment by EgorBo** [comment]
@AndyAyersMS @echesakovMSFT thanks for review, unfortunately I caught a small regression in crossgen (and crossgen2) so I decided to temporarily leave Object.GetType to be old-style.
Now there are zero size regressions in PMI, crossgen and crossgen2 for corelib and framework libs.

**Comment by EgorBo** [comment]
So does anybody mind if I merge it (it contains JIT-EE id update)? It's approved and has zero regression across JIT, crossgen and crossgen2

**Comment by AndyAyersMS** [comment]
Can you say a bit more about the problems you had with converting ` Object.GetType` and why you ultimately decided you could live without it...perhaps in an issue you open to track converting it to new style?

File: src/coreclr/vm/interpreter.cpp [hunk 3]
```diff
@@ -9036,18 +9036,6 @@ void Interpreter::DoCallWork(bool virtualCall, void* thisArg, CORINFO_RESOLVED_T
     {
         switch (intrinsicId)
         {
-        case CORINFO_INTRINSIC_StringLength:
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/vm/interpreter.cpp:9039]: Oh, I thought it's `public int Length => checked((int)Unsafe.As<RawArrayData>(this).Length);` just like in `Array.CoreCLR.cs`
I wonder if I can get rid of the InternalCall in string too

**Comment by EgorBo** [comment]
> Can you say a bit more about the problems you had with converting ` Object.GetType` and why you ultimately decided you could live without it...perhaps in an issue you open to track converting it to new style?

I didn't give up, just wanted to do it iteratively for all of them eventually to keep the PRs easier to review.

The primary goal is the ability to recognize `typeof(T) cmp typeof(T)` in inliner (and `String.Length`), from my understanding - foldable `obj.GetType() == typeof()` patterns where obj's type is known/final are less common (but I might be wrong).

`Object.GetType` led to this assert: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/importer.cpp#L8185

`eeCallInfo` returned `CORINFO_VIRTUALCALL_STUB` for it instead. I suspect the problem is here: https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs#L1587-L1589 (and a similar code in the native crossgen1)

Both crossgens know nothing about the new style intrinsics, I tried to fix this specific case like this https://github.com/dotnet/runtime/pull/52156/commits/65a9154a5d4a2d8518022c58535d8a5a0c169872 and looked like it helped but none-windows platforms still asserted so I decided to postpone it a bit 🙁

File: src/coreclr/vm/interpreter.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/vm/interpreter.cpp:9039]: It looks good to me with a nit. You can also move both of the String methods you have touched to shared CoreLib partition.

**Comment by AndyAyersMS** [comment]
It can be separate. Open an issue and record the details and link it to this PR.

**Comment by EgorBo** [comment]
> It can be separate. Open an issue and record the details and link it to this PR.

Done: https://github.com/dotnet/runtime/issues/52620

**Pull request merged**